tests/test_declaration.py

```diff
@@ -434,6 +434,47 @@ def test_property_with_choices(self):
         self.assertIs(node.style.display, INLINE)
         self.assertTrue(node.style.dirty)
 
+    def test_property_with_storage_class(self):
+        node = TestNode(style=CSS())
+        node.layout.dirty = None
+
+        # Initial value
+        self.assertEqual(repr(node.style.border_spacing), 'BorderSpacing(0px)')
+        self.assertEqual(str(node.style.border_spacing), '0px')
+
+        # Text value
+        node.style.border_spacing = 'inherit'
+        self.assertEqual(repr(node.style.border_spacing), 'BorderSpacing("inherit")')
+        self.assertEqual(str(node.style.border_spacing), 'inherit')
+
+        # Tuple
+        node.style.border_spacing = (1, 2)
+        self.assertEqual(repr(node.style.border_spacing), 'BorderSpacing(1px, 2px)')
+        self.assertEqual(str(node.style.border_spacing), '1px 2px')
+        self.assertEqual(node.style.border_spacing.vertical, 2 * px)
+        self.assertEqual(node.style.border_spacing.horizontal, 1 * px)
+
+        # String with two values
+        node.style.border_spacing = '1 2'
+        self.assertEqual(repr(node.style.border_spacing), 'BorderSpacing(1px, 2px)')
+
+        # String with one value
+        node.style.border_spacing = '1'
+        self.assertEqual(repr(node.style.border_spacing), 'BorderSpacing(1px)')
+        self.assertEqual(str(node.style.border_spacing), '1px')
+        self.assertEqual(node.style.border_spacing.vertical, 1 * px)
+        self.assertEqual(node.style.border_spacing.horizontal, 1 * px)
+
+        # Check error
+        with self.assertRaises(ValueError):
+            node.style.border_spacing = 'a'
+
+        with self.assertRaises(ValueError):
+            node.style.border_spacing = '1 2 3'
+
+        with self.assertRaises(ValueError):
+            node.style.border_spacing = 1, 2, 3
```

Comment:
When you have a test that has internal sections, it's generally a good idea to ask whether it's a separate test. Multiple distinct tests, each of which asserts a single property/feature are almost always preferable to a single monolithic test. The only notable exception is when there's a significant complication to setting up the test in the first place, or if verifying the second feature requires re-running the first test (e.g., testing that an undo reverses the state of the initial action).

Reply:
Added more tests!

Reply:
Yeah... but still not enough. Every time you're putting a `# check special property works` comment in a test - *thats a new test*. And even of the assertions inside those groupings are distinct tests.

For example - "# Invalid amount of items" is *three* tests:
1) Does it raise an error with no input
2) Does it raise an error with three items, in a string, separated with spaces
3) Does it raise an error with three items, separated with commas.

Reply:
Fixed!